### ISSUE TITLE
[BUG]: Live Preview onyl works on first colage photo

### ISSUE THREAD

**Issue by srvoets** [PhotoboothProject/photobooth#942]
### Describe the bug

Live preview loads for first collage photo, then during countdown for second photo, the "preview" of the first image remains.

### Steps to reproduce the behavior

Take collage photo with live preview

### Expected behavior

I should get a live preview before each photo during countdown

### Photobooth Version

4.99.0

### Did it work before?

Yes

### Latest working version

I dont recall, it was in august when I tested last.

**Comment by lattaai2** [comment]
Hello, I tried to fix the issue

This is what I did:
I've identified the issue in the `core.js` file. The problem occurs because the preview is not being reset between collage photos. To fix this, I've modified the `api.thrill` function to reset the preview before each collage photo is taken. Here are the changes I made:

I also created pull request: https://github.com/PhotoboothProject/photobooth/pull/944


> [!CAUTION]  
> Disclaimer: This fix was created by Latta AI and you should never merge before you check the correctness of generated code!
> 
> The fix provided by Latta AI might not be complete and it can serve as an inspiration.

---
This bug was fixed for free by Latta AI - https://latta.ai/ourmission

If you no longer want Latta AI to attempt fixing issues on your repository, you can block this account.

**Comment by nbosb** [comment]
I think this should be fixed by #1267, could you can test it again @srvoets?

**Comment by andi34** [comment]
Since there's a poor description (not mentioned which preview setup is used) of the bug it could also be the problem fixed here: https://github.com/PhotoboothProject/photobooth/pull/1136

**Comment by reloxx13** [comment]
> Since there's a poor description (not mentioned which preview setup is used) of the bug it could also be the problem fixed here: [#1136](https://github.com/PhotoboothProject/photobooth/pull/1136)

nope, problem dod not get fixed with this



> I think this should be fixed by [#1267](https://github.com/PhotoboothProject/photobooth/pull/1267), could you can test it again [@srvoets](https://github.com/srvoets)?

but finally with this, collage is usable again after 1 year.

**Comment by reloxx13** [comment]
Hey,
with this change, the preview is not loading on ipad 9. gen ios 17.5.1, tested in safari and firefox. works fine on windows firefox.

i rolled back to an backup before this change its working again

**Comment by andi34** [comment]
@reloxx13 updated the URL config? The `url()` must be removed.

Myself has no hands to an iOS device, can never guarantee compatibility.

**Comment by reloxx13** [comment]
Hey, yes, the config was without url().

The failure seems specific to the infinite multipart/x-mixed-replace and Transfer-Encoding: chunked combination, which Safari WebKit doesnt like. I just did some research.

**Comment by reloxx13** [comment]
I guess easiest way to fix would be to use the addCacheBustingParam on the background-image.

**Comment by andi34** [comment]
Could you try 
http://localhost:1984/api/stream.m3u8?src=photobooth
Or
http://localhost:1984/api/webrtc.html?src=photobooth

As url config?

**Comment by reloxx13** [comment]
> Could you try http://localhost:1984/api/stream.m3u8?src=photobooth Or http://localhost:1984/api/webrtc.html?src=photobooth
> 
> As url config?

neither worked

to get it fixed i added this to the old version with the div, not img.


```
    function addCacheBustingParam(url) {
        const timestamp = new Date().getTime();

        if (url.includes('?')) {
            return `${url}&t=${timestamp}`;
        }

        return `${url}?t=${timestamp}`;
    }

    function getRootProperty(property) {
        const root = document.documentElement;
        const style = getComputedStyle(root);
        return style.getPropertyValue(property).trim();
    }
    
    function parseCssUrl(cssValue) {
        const match = cssValue.match(/url\(["']?(.*?)["']?\)/);
        return match ? match[1] : null;
    }

....

    photoboothTools.console.logDev('Preview: Preview at countdown from URL.');
    const cssValue = getRootProperty('--background-preview'); //get orginal url
    const baseUrl = parseCssUrl(cssValue); // remove url("")
    const cacheBusted = addCacheBustingParam(baseUrl); // add timestampe
    url.css('background-image', `url("${cacheBusted}")`);
    setTimeout(function () {

.....

```

**Comment by reloxx13** [comment]
how do we continue with this?

**Comment by reloxx13** [comment]
@nbosb what is your streaming URL? it does not work on my ipad as explained above
mine is `http://192.168.4.1:1984/api/stream.mjpeg?src=photobooth` with go2rtc 1.9.7 

```
---
streams:
  photobooth: exec:gphoto2 \
        --set-config capture=on \
        --set-config /main/capturesettings/aperture=4.5 \
        --set-config /main/capturesettings/shutterspeed=1/15 \
        --set-config /main/imgsettings/iso=800 \
        --capture-movie --stdout#killsignal=2

log:
  exec: trace

```

**Comment by nbosb** [comment]
I'm using the same configuration (go2rtc 1.9.7 + MJPEG stream) and the same URL as you (just with a different IP). I'm testing on an iPad 6th Gen running iOS 17.7.10. I tried your suggested fix of switching from an `<img>` element back to a `<div>`. I've committed the changes to a branch: [https://github.com/nbosb/photobooth/tree/dev-preview](https://github.com/nbosb/photobooth/tree/dev-preview).

However, with this change, I’m encountering an issue where the stream is still being fetched even after the preview is hidden. As a result, the camera isn’t being released by go2rtc, and the picture-taking process fails.

**Comment by nbosb** [comment]
My fault, my capture command was broken. I could fix my issue by using the `/usr/local/bin/capture` script that stops go2rtc.

The `<div>` element works for me, too.

**Comment by reloxx13** [comment]
> My fault, my capture command was broken. I could fix my issue by using the `/usr/local/bin/capture` script that stops go2rtc.
> 
> The `<div>` element works for me, too.

heya,
nono, im having the same issues and i used the capture command. check my comments here in the core.js file:
https://github.com/PhotoboothProject/photobooth/commit/b25baa29ee714a9da02011e6e9b976291a897c7b#r171135714

<img width="807" height="857" alt="Image" src="https://github.com/user-attachments/assets/a9c9b575-692a-428a-ac19-7cd023a1d082" />

Im on it the 2nd day now trying to find a working solution and just found this timing issue.
do you have this stop preview time before capture set in config? or do you have something in the kill command input?

**Comment by nbosb** [comment]
No, I have set the _Time before the countdown ends to execute the preview stop command_ to zero. I'm using the capture script that adds a 0.5s delay between stopping go2rtc and taking the picture as described here: https://photoboothproject.github.io/faq/go2rtc-troubleshooting/#delay-capture

**Comment by reloxx13** [comment]
> No, I have set the _Time before the countdown ends to execute the preview stop command_ to zero. I'm using the capture script that adds a 0.5s delay between stopping go2rtc and taking the picture as described here: https://photoboothproject.github.io/faq/go2rtc-troubleshooting/#delay-capture

yes i have this, too.
i just think my cam is slower as your because i need more that 0.5s pause. canon 600d

**Comment by reloxx13** [comment]
and i dont have a cheese time which is also important for the timings: countdown, (if stop time: stop at time) countdown done, stop preview, show cheese, take pic, rt2rtc stop, sleep 0,5, capture (optional wait event 300ms), start rt2rtc.

without cheese, stop time and sleep in alternative capture, there is no pause between preview and capture, blocking the cam

**Comment by andi34** [comment]
https://photoboothproject.github.io/faq/go2rtc-troubleshooting/

**Comment by reloxx13** [comment]
> https://photoboothproject.github.io/faq/go2rtc-troubleshooting/

everything worked fine before this commit or before adding cache busting.
i have a working backup with a installation of 9.10.2025.

the troubleshooting is not helpfull in this case.

**Comment by andi34** [comment]
Have you tried an updated go2rtc? v1.9.12 is available via go2rtc Installer.

https://github.com/PhotoboothProject/photobooth/commit/0f31f595da89871d250ed62bf06a3bbcdb533abc

**Comment by andi34** [comment]
Another URL to try:
```
http://192.168.4.1:1984/stream.mjpeg?src=photobooth
```

You could also play with the go2rtc config further (/etc/go2rtc.yaml)
```
api:
  mjpeg:
    write_buffer: 0
    read_buffer: 0
```

@nbosb from what I've read MJPEG via `<img>` or `<video>` will fail on Safari and maybe other mobile Browser on Android too. We could consider adding some JavaScript and use webrtc, but currently I've not really time to look at it further.

**Comment by reloxx13** [comment]
> Have you tried an updated go2rtc? v1.9.12 is available via go2rtc Installer.
> 
> [0f31f59](https://github.com/PhotoboothProject/photobooth/commit/0f31f595da89871d250ed62bf06a3bbcdb533abc)

yes, tried latest

**Comment by reloxx13** [comment]
> Another URL to try:
> 
> ```
> http://192.168.4.1:1984/stream.mjpeg?src=photobooth
> ```
404 not found

> You could also play with the go2rtc config further (/etc/go2rtc.yaml)
> 
> ```
> api:
>   mjpeg:
>     write_buffer: 0
>     read_buffer: 0
> ```
will try...


> [@nbosb](https://github.com/nbosb) from what I've read MJPEG via `<img>` or `<video>` will fail on Safari and maybe other mobile Browser on Android too. We could consider adding some JavaScript and use webrtc, but currently I've not really time to look at it further.


safari does not allow autoplay webrtc cause of invasive ads x.x
edit: only if not muted afaik, sorry

**Comment by andi34** [comment]
Can't tell to be true and no hands to an iOS device and no money left to buy one for development purposes.
Keeping iOS compatible is always part of to the community work. If you've got any working changes let me know and I am glad to review.

**Comment by reloxx13** [comment]
I have implemented WebRTC, but it was too laggy and slow.
ordered an android tablet now.

**Comment by reloxx13** [comment]
so the img tag does not work on my ipad, i dont know why it is working on yours. are you sure you have the latest code on your hosting device?

i changed it back and added cacheBusting to the div background-image and this is working.

But the latest go2rtc version 1.9.12 breaks the preview. the cam is stucking in busy or complete dark lcd when switching from video to capture.
I tried everything from the troubleshooting, trying around with the timings (up to 3seconfs sleep or wait event time) and spent alot of time getting the cache busting working and in the end it was go2rtc. 

im back on 1.9.7 and finally the cam and preview is working again. im trying to find the exact breaking version now.

**Comment by reloxx13** [comment]
go2rtc version 1.9.9 and higher is breaking and blocking my cam

**Comment by reloxx13** [comment]
@nbosb did you have time to test the new pr? why does it work for you and for me not if we are on the same device?

**Comment by nbosb** [comment]
Sorry for the late reply, I've only just found the time to look at the PR. The current dev branch (which contains #1302) is working fine for me. The preview runs, it is working with collages and it runs on the iPad.
